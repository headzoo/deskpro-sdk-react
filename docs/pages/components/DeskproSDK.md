Overview
========
The `DeskproSDK` component bootstraps the app and [connects it to the SDK](/pages/props/#connecting-your-components).

```jsx
import ReactDOM from 'react-dom';
import { DeskproSDK, configureStore } from 'deskpro-sdk-react';

const store = configureStore(dpapp);

export function runApp(dpapp) {
  ReactDOM.render(
    <DeskproSDK dpapp={dpapp} store={store}>
      <App />
    </DeskproSDK>,
    document.getElementById('deskpro-app')
  );
}

```

The app component may also be passed to `DeskproSDK` via the `component` prop.

```jsx
import ReactDOM from 'react-dom';
import { DeskproSDK, configureStore } from 'deskpro-sdk-react';

const store = configureStore(dpapp);

export function runApp(dpapp) {
  ReactDOM.render(
    <DeskproSDK dpapp={dpapp} store={store} component={App} />,
    document.getElementById('deskpro-app')
  );
}

```

## Props

```jsx
<DeskproSDK
    dpapp={object}
    store={object}
    component={element}
/>
```
