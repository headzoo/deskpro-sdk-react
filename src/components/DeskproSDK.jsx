import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { AppEvents, UIEvents } from '@deskproapps/deskproapps-sdk-core';
import { Loader, Alert, DrawerList } from 'deskpro-components';
import { dpappPropType, storePropType } from '../utils/props';
import * as sdkActions from '../actions/sdkActions';
import { sdkProps } from '../utils/connect';
import Route from '../utils/route';

/**
 * Connects DeskPRO apps to the DeskPRO API
 *
 * Provides SDK props to a wrapped component. The props provide information about the app
 * and have methods which communicate with the API.
 *
 * Example:
 *
 * ```jsx
 * import ReactDOM from 'react-dom';
 * import { DeskproSDK, configureStore } from 'deskpro-sdk-react';
 * import App from './App';
 *
 * export function runApp(dpapp) {
 *  const store = configureStore(dpapp);
 *
 *  ReactDOM.render(
 *    <DeskproSDK dpapp={dpapp} store={store}>
 *      <App />
 *    </DeskproSDK>,
 *    document.getElementById('deskpro-app')
 *  );
 * }
 * ```
 */
class DeskproSDK extends React.Component {
  static propTypes = {
    /**
     * Instance of sdk-core.
     */
    dpapp: dpappPropType.isRequired,

    /**
     * Instance of the redux store.
     */
    store: storePropType.isRequired,

    /**
     * The sdk values stored in the store.
     */
    sdk: PropTypes.object.isRequired,

    /**
     * Bound action creators.
     */
    actions: PropTypes.object.isRequired,

    /**
     * The app component.
     */
    component: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),

    /**
     * The app component.
     */
    children: PropTypes.element
  };

  /**
   * Specifies the default values for props
   */
  static defaultProps = {
    component: null,
    children:  null
  };

  /**
   * Context values passed down to children that declare contextTypes
   */
  static childContextTypes = {
    dpapp: dpappPropType,
    store: storePropType,
    route: PropTypes.object
  };

  /**
   * @returns {{dpapp: *}}
   */
  getChildContext = () => {
    const { dpapp, store } = this.props;

    return {
      route: new Route(store.dispatch, store.getState().sdk.route),
      store,
      dpapp
    };
  };

  /**
   * Bootstraps the application
   */
  componentDidMount = () => {
    const { dpapp } = this.props;

    this.bootstrap();
    dpapp.on(AppEvents.EVENT_REFRESH, this.handleRefresh);
    dpapp.on(UIEvents.EVENT_UI_DISPLAYCHANGED, this.handleDisplayChange);
  };

  /**
   * Triggered before the component is unmounted
   */
  componentWillUnmount = () => {
    const { dpapp } = this.props;

    dpapp.off(AppEvents.EVENT_REFRESH, this.handleRefresh);
    dpapp.off(UIEvents.EVENT_UI_DISPLAYCHANGED, this.handleDisplayChange);
  };

  /**
   * Bootstraps the app component
   *
   * @returns {Promise}
   */
  bootstrap = () => {
    const { actions } = this.props;

    actions.ready(false);
    return Promise.all([
      ...this.bootstrapMe(),
      ...this.bootstrapTabData(),
      ...this.bootstrapStorage()
    ])
      .then(() => {
        return actions.ready();
      })
      .catch((error) => {
        return actions.error(error);
      });
  };

  /**
   * Fetches the "me" data for the user
   *
   * @returns {Promise[]}
   */
  bootstrapMe = () => {
    const { dpapp, actions } = this.props;

    const promise = dpapp.restApi.get('/me')
      .then((resp) => {
        try {
          return Promise.resolve(actions.me(resp.body.data.person));
        } catch (e) {
          return Promise.resolve({});
        }
      });

    return [promise];
  };

  /**
   * Fetches the data for the active tab
   *
   * @returns {Promise[]}
   */
  bootstrapTabData = () => {
    const { dpapp, actions } = this.props;

    const promise = dpapp.context.getTabData()
      .then((resp) => {
        try {
          return Promise.resolve(actions.tabData(resp.api_data));
        } catch (e) {
          return Promise.resolve({});
        }
      });

    return [promise];
  };

  /**
   * Fetches the manifest storage values
   *
   * @returns {Promise[]}
   */
  bootstrapStorage = () => {
    const { dpapp, actions } = this.props;

    const promises = [];
    const items = dpapp.manifest.storage || dpapp.manifest.state;
    if (items && items.length > 0) {
      const appKeys    = [];
      const entityKeys = [];
      const oauthKeys  = [];

      items.forEach((item) => {
        if (item.name.indexOf('oauth:') === 0) {
          oauthKeys.push(
            item.name.replace('oauth:', '')
          );
        } else if (item.name.indexOf('entity:') === 0) {
          entityKeys.push(
            item.name.replace('entity:', '')
          );
        } else {
          appKeys.push(
            item.name.replace('app:', '')
          );
        }
      });

      if (appKeys.length > 0) {
        promises.push(actions.appGetStorage(appKeys));
      }
      if (entityKeys.length > 0) {
        promises.push(actions.entityGetStorage(entityKeys));
      }
      oauthKeys.forEach((key) => {
        promises.push(actions.oauthGetSettings(key));
      });
    }

    return promises;
  };

  /**
   * Callback for the AppEvents.EVENT_REFRESH event
   */
  handleRefresh = () => {
    const { actions } = this.props;

    actions.refreshing(true);
    this.bootstrap()
      .catch((error) => {
        return actions.error(error);
      })
      .then(() => actions.refreshing(false));
  };

  /**
   * Callback for the UIEvents.EVENT_UI_DISPLAYCHANGED event
   */
  handleDisplayChange = () => {
    const { actions, sdk } = this.props;

    actions.collapsed(!sdk.ui.collapsed);
  };

  /**
   * Renders SDK errors as alerts
   *
   * @returns {Array}
   */
  renderErrors = () => {
    return this.props.sdk.errors.map(error => (
      <Alert key={error.id} type="danger" style={{ margin: 10 }}>
        {error.msg}
      </Alert>
    ));
  };

  /**
   * Renders the loading animation
   *
   * @returns {XML}
   */
  renderLoading = () => {
    return (
      <div className="dp-text-center">
        <Loader />
      </div>
    );
  };

  /**
   * Renders the main app component
   *
   * @returns {XML}
   */
  renderApp = () => {
    const { component, children } = this.props;

    if (!component && !children) {
      throw new Error('App component not set.');
    }

    return React.cloneElement(
      component || React.Children.only(children),
      sdkProps(this.props)
    );
  };

  /**
   * @returns {XML}
   */
  render() {
    const { sdk } = this.props;

    return (
      <DrawerList>
        <li
          className="dp-column-drawer--with-controls"
          style={{ display: sdk.ui.collapsed ? 'none' : 'block' }}
        >
          {this.renderErrors()}
          {!sdk.ready || sdk.ui.loading
            ? this.renderLoading()
            : this.renderApp()
          }
        </li>
      </DrawerList>
    );
  }
}

/**
 * Maps redux state to component props
 *
 * @param {*} state
 * @returns {{sdk: *}}
 */
function mapStateToProps(state) {
  return {
    sdk: Object.assign({}, state.sdk)
  };
}

/**
 * Maps action creators to component props
 *
 * @param {object} dispatch
 * @returns {{actions: *}}
 */
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(sdkActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(DeskproSDK);
