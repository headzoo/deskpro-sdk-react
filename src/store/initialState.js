export default {
  form: {},
  sdk:  {
    me:      {},
    tabData: {},
    errors:  [],
    ready:   false,
    oauth:   {
      providers: {}
    },
    ui: {
      loading:    false,
      collapsed:  false,
      refreshing: false
    },
    route: {
      location: 'index',
      params:   {}
    },
    storage: {
      app:    {},
      entity: {}
    }
  }
};
