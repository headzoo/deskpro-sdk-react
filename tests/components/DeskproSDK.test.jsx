import React from 'react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import initialState from '../../src/store/initialState';
import DeskproSDK from '../../src/components/DeskproSDK';

const mockStore = configureMockStore();
const dpapp = {
  context: {
    getTabData: () => Promise.resolve()
  },
  on:      () => {},
  oauth:   {},
  ui:      {},
  restApi: {
    get: () => Promise.resolve()
  },
  manifest: {
    storage: []
  }
};

it('DeskproSDK renders the wrapped component', () => {
  const App = () => <div>App</div>;
  const store = mockStore(initialState);
  const renderedValue = renderer.create(
    <DeskproSDK dpapp={dpapp} store={store}>
      <App />
    </DeskproSDK>
  ).toJSON();

  expect(renderedValue).toMatchSnapshot();
});

it('DeskproSDK renders the prop component', () => {
  const App = () => <div>App</div>;
  const store = mockStore(initialState);
  const renderedValue = renderer.create(
    <DeskproSDK dpapp={dpapp} store={store} component={App} />
  ).toJSON();

  expect(renderedValue).toMatchSnapshot();
});
