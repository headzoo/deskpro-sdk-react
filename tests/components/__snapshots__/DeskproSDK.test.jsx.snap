// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeskproSDK renders the prop component 1`] = `
<ul
  className="dp-list dp-column-drawer-list"
  role="tree"
>
  <li
    className="dp-column-drawer--with-controls"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      className="dp-text-center"
    >
      <span
        className="SVGInline dp-loader dp-loader--xl"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg class=\\"SVGInline-svg dp-loader-svg dp-loader--xl-svg\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
<g class=\\"st0\\">
  <path class=\\"st1\\" d=\\"M8 2c3.3 0 6 2.7 6 6s-2.7 6-6 6 -6-2.7-6-6S4.7 2 8 2M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8S12.4 0 8 0L8 0z\\"/>
</g>
<path class=\\"st2\\" d=\\"M8 2V0C3.6 0 0 3.6 0 8h2C2 4.7 4.7 2 8 2z\\"/>
<path class=\\"st3\\" d=\\"M2 8c0-1.7 0.7-3.2 1.8-4.2L2.3 2.3C0.9 3.8 0 5.8 0 8s0.9 4.2 2.3 5.7l1.4-1.4C2.7 11.2 2 9.7 2 8z\\"/>
<path class=\\"st1\\" d=\\"M8 0v2c3.3 0 6 2.7 6 6s-2.7 6-6 6c-1.7 0-3.2-0.7-4.2-1.8l-1.4 1.4C3.8 15.1 5.8 16 8 16c4.4 0 8-3.6 8-8C16 3.6 12.4 0 8 0z\\"/>
</svg>",
          }
        }
      />
    </div>
  </li>
</ul>
`;

exports[`DeskproSDK renders the wrapped component 1`] = `
<ul
  className="dp-list dp-column-drawer-list"
  role="tree"
>
  <li
    className="dp-column-drawer--with-controls"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      className="dp-text-center"
    >
      <span
        className="SVGInline dp-loader dp-loader--xl"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg class=\\"SVGInline-svg dp-loader-svg dp-loader--xl-svg\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\">
<g class=\\"st0\\">
  <path class=\\"st1\\" d=\\"M8 2c3.3 0 6 2.7 6 6s-2.7 6-6 6 -6-2.7-6-6S4.7 2 8 2M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8S12.4 0 8 0L8 0z\\"/>
</g>
<path class=\\"st2\\" d=\\"M8 2V0C3.6 0 0 3.6 0 8h2C2 4.7 4.7 2 8 2z\\"/>
<path class=\\"st3\\" d=\\"M2 8c0-1.7 0.7-3.2 1.8-4.2L2.3 2.3C0.9 3.8 0 5.8 0 8s0.9 4.2 2.3 5.7l1.4-1.4C2.7 11.2 2 9.7 2 8z\\"/>
<path class=\\"st1\\" d=\\"M8 0v2c3.3 0 6 2.7 6 6s-2.7 6-6 6c-1.7 0-3.2-0.7-4.2-1.8l-1.4 1.4C3.8 15.1 5.8 16 8 16c4.4 0 8-3.6 8-8C16 3.6 12.4 0 8 0z\\"/>
</svg>",
          }
        }
      />
    </div>
  </li>
</ul>
`;
